test/widgets/compose_box_test.dart

```diff
@@ -195,10 +206,163 @@ void main() {
+        accountId: account.id, page: const ComposeBox(narrow: narrow))));
+      await tester.pumpAndSettle();
+
+      store = await testBinding.globalStore.perAccount(account.id);
+      connection = store.connection as FakeApiConnection;
+    }
+
+    testWidgets('navigating away sends "stopped typing" notice', (tester) async {
+      await prepareComposeBoxWithNavigation(tester);
+
+      await checkStartTyping(tester, narrow);
+
+      connection.prepare(json: {});
+      (await ZulipApp.navigator).pop();
+      await tester.pump(Duration.zero);
+      checkTypingRequest(TypingOp.stop, narrow);
+    });
+
+    testWidgets('for content input, unfocusing sends "stopped typing" notice '
+                'and refocusing does not send "started typing notice', (tester) async {
+      const narrow = ChannelNarrow(123);
+      final destinationNarrow = TopicNarrow(narrow.streamId, 'test topic');
+      await prepareComposeBox(tester, narrow: narrow, account: account);
+
+      await tester.enterText(topicInputFinder, destinationNarrow.topic);
+      // Remove an irrelevant topic request.
+      check(connection.takeRequests()).single
+        ..method.equals('GET')
+        ..url.path.equals('/api/v1/users/me/123/topics');
+
+      connection.prepare(json: {});
+      await tester.enterText(contentInputFinder, 'hello world');
+      checkTypingRequest(TypingOp.start, destinationNarrow);
+
+      connection.prepare(json: {});
+      // Move focus to the topic input.
+      await tester.tap(topicInputFinder);
+      await tester.pump(Duration.zero);
+      checkTypingRequest(TypingOp.stop, destinationNarrow);
+
+      await tester.tap(contentInputFinder);
+      await tester.pump(Duration.zero);
+      // Refocusing on the content input triggers an update to
+      // [TextEditingController.selection].  Still, we expect no
+      // "typing started" request because the text remains the same.
+    });
+
+    testWidgets('unfocusing app sends "typing stopped" notice', (tester) async {
+      Future<void> setAppLifeCycleState(AppLifecycleState state) async {
+        // While this state lives on [ServicesBinding], testWidgets resets it
+        // for us when the test ends:
+        //   https://github.com/flutter/flutter/blob/c78c166e3ecf963ca29ed503e710fd3c71eda5c9/packages/flutter_test/lib/src/binding.dart#L1189
+        final ByteData? message = const StringCodec().encodeMessage(state.toString());
+        await tester.binding.defaultBinaryMessenger
+          .handlePlatformMessage('flutter/lifecycle', message, (_) {});
+      }
+
+      await prepareComposeBox(tester, narrow: narrow, account: account);
+
+      await checkStartTyping(tester, narrow);
+
+      connection.prepare(json: {});
+      await setAppLifeCycleState(AppLifecycleState.inactive);
+      checkTypingRequest(TypingOp.stop, narrow);
+      // Wait for the pending timer from the API request to end.
+      await tester.pump(Duration.zero);
+    });
```

Comment:
Huh, why does this test case need this step when the other new test cases don't?

Reply:
When the request is triggered by an `enterText` call or a timeout, this is not needed. In the other cases, `await tester.pump` might have been called a bit earlier for different purposes, which also waits for the API request to end.

Reply:
Ah I see, subtle. I think the point is that when timers get run, that leads to all timers getting run that are due at the ending time — which in particular means all fake API requests that are started at that time (and don't set a positive delay) get completed.

I think it might be less confusing if the last two lines are swapped, so that it ends the same kind of way as the other tests (and exactly the same way as the "navigating away" test, which is the most similar test case to this one):
```dart
      await tester.pump(Duration.zero);
      checkTypingRequest(TypingOp.stop, narrow);
```

That works, for the same reason the others work.